wasi/wasi.go

```diff
@@ -15,6 +15,75 @@ import (
+// wasm.Store Memories index zero: `store.Memories[0].Buffer`
+//
+// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md
+// See https://www.w3.org/TR/wasm-core-1/#memory-instances%E2%91%A0.
+type Api interface {
+	// ArgsSizesGet is a WASI function that reads command-line argument data (Args) sizes.
+	//
+	// There are two parameters and an Errno result. Both parameters are offsets in the wasm.MemoryInstance Buffer to
+	// write the corresponding sizes in uint32 little-endian encoding.
+	//
+	// * argc - is the offset to write the argument count to the wasm.MemoryInstance Buffer
+	// * argvBufSize - is the offset to write the null terminated argument length to the wasm.MemoryInstance Buffer
+	//
+	// In WebAssembly 1.0 (MVP) Text format, this signature is:
+	//	(import "wasi_snapshot_preview1" "args_sizes_get"
+	//		(func ($wasi_args_sizes_get (param $argc i32) (param $argv_buf_size i32) (result i32))
+	//
+	// For example, if Args are []string{"a","bc"} and
+	//   FunctionArgsSizesGet parameters argc=1 and argvBufSize=6, we expect `store.Memories[0].Buffer` to contain:
+	//
+	//             uint32         uint32
+	//           +--------+     +--------+
+	//           |        |     |        |
+	// []byte{?, 0, 0, 0, 2, ?, 0, 0, 0, 5, ?}
+	//    argc --^        ^     ^        ^
+	//           2 args --+     |        |
+	//            argvBufSize --+        |
+	//     len([]byte{'a',0,'b',c',0}) --+
+	//
+	// See ArgsGet
+	// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#args_sizes_get
+	// See https://en.wikipedia.org/wiki/Null-terminated_string
+	ArgsSizesGet(ctx *wasm.HostFunctionCallContext, argc, argvBufSize uint32) Errno
+
+	// ArgsGet is a WASI function that reads command-line argument data (Args).
+	//
+	// There are two parameters and an Errno result. Both parameters are offsets in the wasm.MemoryInstance Buffer to
+	// write offsets. These are encoded uint32 little-endian.
+	//
+	// * argv - is the offset to begin writing argument offsets to the wasm.MemoryInstance
+	//   * FunctionArgsSizesGet argc * 4 bytes are written to this offset
+	// * argvBuf - is the offset to write the null terminated arguments to the wasm.MemoryInstance
+	//   * FunctionArgsSizesGet argv_buf_size are written to this offset
+	//
+	// In WebAssembly 1.0 (MVP) Text format, this signature is:
+	//	(import "wasi_snapshot_preview1" "args_get"
+	//		(func ($wasi_args_get (param $argv i32) (param $argv_buf i32) (result i32))
+	//
+	// For example, if FunctionArgsSizesGet wrote argc=2 and argvBufSize=5 for arguments: "a" and "bc"
+	//   and FunctionArgsGet parameters argv=6 and argvBuf=1, we expect `store.Memories[0].Buffer` to contain:
+	//
+	//               argvBufSize                argc * 4
+	//            +----------------+     +--------------------+
+	//            |                |     |                    |
+	// []byte{?, 'a', 0, 'b', 'c', 0, ?, 0, 0, 0, 1, 0, 0, 0, 3, ?}
```

Comment:
ps, later I'll rewrite the unit tests (another PR) to look exactly like this. When there are less than a hundred bytes, it is sometimes best to have tests be literal in expectations vs doing codec internal to test functions

Reply:
I will rewrite them now as this actually shows a problem.. basically this is little-endian and it becomes obvious when tests are literal